File: cards.go
package main

import (
	"math/rand"
)

type CardValue int
type CardType int

const (
  TWO CardValue = iota
  THREE
  FOUR
  FIVE
  SIX
  SEVEN
  EIGHT
  NINE
  TEN
  JACK
  QUEEN
  KING
  ACE
)

func (v CardValue) String() string {
  return [...]string{
  "Ace",
  "Two",
  "Three",
  "Four",
  "Five",
  "Six",
  "Seven",
  "Eight",
  "Nine",
  "Ten",
  "Jack",
  "Queen",
  "King",
  }[v]
}

const (
  SPADE CardType = iota
  CLUB
  DIAMOND
  HEART
)

func (t CardType) String() string {
  return [...]string{
  "Spade",
  "Club",
  "Diamond",
  "Heart",
  }[t]
}

type Card struct {
  Value CardValue
  Type CardType
}

type Deck struct {
  Cards []Card
  Used int
}

func DeckFactory() Deck {
  d := Deck{}

  d.Cards = make([]Card, 0)
  d.Used = 0

  d.Cards = append(d.Cards, Card{
    Value: ACE,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: TWO,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: THREE,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: FOUR,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: FIVE,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: SIX,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: SEVEN,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: EIGHT,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: NINE,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: TEN,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: JACK,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: QUEEN,
    Type: SPADE,
  })
  d.Cards = append(d.Cards, Card{
    Value: KING,
    Type: SPADE,
  })

  d.Cards = append(d.Cards, Card{
    Value: ACE,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: TWO,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: THREE,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: FOUR,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: FIVE,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: SIX,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: SEVEN,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: EIGHT,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: NINE,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: TEN,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: JACK,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: QUEEN,
    Type: CLUB,
  })
  d.Cards = append(d.Cards, Card{
    Value: KING,
    Type: CLUB,
  })

  d.Cards = append(d.Cards, Card{
    Value: ACE,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: TWO,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: THREE,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: FOUR,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: FIVE,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: SIX,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: SEVEN,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: EIGHT,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: NINE,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: TEN,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: JACK,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: QUEEN,
    Type: DIAMOND,
  })
  d.Cards = append(d.Cards, Card{
    Value: KING,
    Type: DIAMOND,
  })

  d.Cards = append(d.Cards, Card{
    Value: ACE,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: TWO,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: THREE,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: FOUR,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: FIVE,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: SIX,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: SEVEN,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: EIGHT,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: NINE,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: TEN,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: JACK,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: QUEEN,
    Type: HEART,
  })
  d.Cards = append(d.Cards, Card{
    Value: KING,
    Type: HEART,
  })

  return d
}

func (d Deck) Shuffle() {
  for i := 0; i < len(d.Cards) - 2; i++ {
    random := i + rand.Intn(52 - i)

    temp := d.Cards[i]
    d.Cards[i] = d.Cards[random]
    d.Cards[random] = temp
  }
}

func (d *Deck) Deal() Card {
  d.Used += 1
  return d.Cards[d.Used - 1]
}

File: src/evaluator/cards/cards.zig
pub const constants = @import("./constants.zig");

pub const Suit = enum(u2) {
    club = 0,
    diamond,
    heart,
    spade,
    pub fn name(self: Suit) []const u8 {
        return switch (self) {
            Suit.club => "Clubs",
            Suit.diamond => "Diamonds",
            Suit.heart => "Hearts",
            Suit.spade => "Spades",
        };
    }
};

pub const Rank = enum(u4) {
    ace = 12,
    two = 0,
    three = 1,
    four = 2,
    five = 3,
    six = 4,
    seven = 5,
    eight = 6,
    nine = 7,
    ten = 8,
    jack = 9,
    queen = 10,
    king = 11,
    pub fn name(self: Rank) []const u8 {
        return switch (self) {
            Rank.ace => "Ace",
            Rank.two => "Two",
            Rank.three => "Three",
            Rank.four => "Four",
            Rank.five => "Five",
            Rank.six => "Six",
            Rank.seven => "Seven",
            Rank.eight => "Eight",
            Rank.nine => "Nine",
            Rank.ten => "Ten",
            Rank.jack => "Jack",
            Rank.queen => "Queen",
            Rank.king => "King",
        };
    }
};

fn get_suit_mask(suit: Suit) u32 {
    return switch (suit) {
        .club => 0x00008000,
        .diamond => 0x00004000,
        .heart => 0x00002000,
        .spade => 0x00001000,
    };
}

fn get_rank_mask(rank: Rank) u32 {
    return switch (rank) {
        .ace => 0x10000C29,
        .king => 0x08000B25,
        .queen => 0x04000A1F,
        .jack => 0x0200091D,
        .ten => 0x01000817,
        .nine => 0x00800713,
        .eight => 0x00400611,
        .seven => 0x0020050D,
        .six => 0x0010040B,
        .five => 0x00080407,
        .four => 0x00040205,
        .three => 0x00200103,
        .two => 0x00010002,
    };
}

pub const Card = struct {
    Suit: Suit,
    Rank: Rank,

    Mask: u32,

    pub fn get_suit(self: Card) Suit {
        _ = self;
    }

    pub fn get_prime(self: Card) usize {
        return self.Mask & 0xFF;
    }
};

pub fn CardFactory(rank: Rank, suit: Suit) Card {
    return Card{ .Rank = rank, .Suit = suit, .Mask = get_rank_mask(rank) | get_suit_mask(suit) };
}
